Updates project to use cloudflare-go

KAllan357 opened the pull request:
Hi!

This PR should address a few outstanding issues in the repo. Specifically, https://github.com/adyanth/cloudflare-operator/issues/76, https://github.com/adyanth/cloudflare-operator/issues/7, and https://github.com/adyanth/cloudflare-operator/issues/47.

After executing a pass against a few of the functions  inside `cloudflare_api.go`, I decided that this could go all the way and essentially rewrite the whole file. My high level goal was to keep the API consistent, but adjust the functions so that they used `cloudflare-go` instead of the `net/http` package.

Please take a look when you get a chance.

KAllan357 commented:
@adyanth tagging you to see if you'd consider / review this PR.

adyanth commented:
Hey @KAllan357 wow, thank you very much for the contribution! I was busy the last week, did not see this. Will review.

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -104,46 +64,35 @@ func (c *CloudflareAPI) CreateCloudflareTunnel() (string, string, error) {
-	creds, _ := json.Marshal(tunnelResponse.Result.CredentialsFile)
-	return tunnelResponse.Result.Id, string(creds), nil
+	// Marshal the tunnel credentials into a string
+	creds, _ := json.Marshal(credentialsFile)
```
Store and return the error for Marshal in the next line

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -153,27 +102,20 @@ func (c *CloudflareAPI) DeleteCloudflareTunnel() error {
-	if err := c.addAuthHeader(req, true); err != nil {
-		return err
-	}
+	ctx := context.Background()
```
We do have a context from the Reconcile method, should we use the same?

Edit: It would need to pass through 3 different functions. I will make a separate PR for that actually.

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -324,28 +247,16 @@ func (c *CloudflareAPI) validateTunnelId() bool {
+	tunnel, err := c.CloudflareClient.GetTunnel(ctx, rc, c.TunnelId)
 	if err != nil {
-		c.Log.Error(err, "error code in getting tunnel by Tunnel ID", "tunnelId", c.TunnelId)
+		c.Log.Error(err, "error retrieving tunnel")
```
Any reason to remove the tunnel ID from the log? I thought it might be easier to tie what errors point to what tunnels since it would not have any identifier otherwise.

adyanth replied:
Any reason to remove the tunnel ID from the log? I thought it might be easier to tie what errors point to what tunnels since it would not have any identifier otherwise.
```suggestion
		c.Log.Error(err, "error retrieving tunnel", "tunnelId", c.TunnelId)
```

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -431,96 +335,72 @@ func (c *CloudflareAPI) GetZoneId() (string, error) {
 	}
 }
 
+// Been a while writing Go... returns a pointer to a type
```
Wow, that is a very nice way to return a pointer to an untyped constant. Never seen that before. Will remember this :)

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -2,69 +2,45 @@ package controllers
-const CLOUDFLARE_ENDPOINT = "https://api.cloudflare.com/client/v4/"
-
 // TXT_PREFIX is the prefix added to TXT records for whom the corresponding DNS records are managed by the operator.
 const TXT_PREFIX = "_managed."
```
I do not think this is needed anymore since the TXT is being added on the CNAME itself, right?

adyanth reviewed (changes requested):
Excellent! Left a couple comments on tiny nit picks :)

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -104,46 +64,35 @@ func (c *CloudflareAPI) CreateCloudflareTunnel() (string, string, error) {
-	creds, _ := json.Marshal(tunnelResponse.Result.CredentialsFile)
-	return tunnelResponse.Result.Id, string(creds), nil
+	// Marshal the tunnel credentials into a string
+	creds, _ := json.Marshal(credentialsFile)
+	return tunnel.ID, string(creds), nil
```
Store and return the error for Marshal in the next line?
```suggestion
	creds, err := json.Marshal(credentialsFile)
	return tunnel.ID, string(creds), err
```

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -2,69 +2,45 @@ package controllers
-// CLOUDFLARE_ENDPOINT is the Cloudflare API base URL from https://api.cloudflare.com/#getting-started-endpoints.
-const CLOUDFLARE_ENDPOINT = "https://api.cloudflare.com/client/v4/"
-
 // TXT_PREFIX is the prefix added to TXT records for whom the corresponding DNS records are managed by the operator.
 const TXT_PREFIX = "_managed."
 
```
I do not think this is needed anymore since the TXT is being added on the CNAME itself, right? 
```suggestion
```

KAllan357 replied:
I agree that upon inspection, its not used anymore. I'm not sure the TXT record will have that prefix without this though, it might just be the same as `fqdn`. Probably need to test this, but  can remove it for now.

adyanth replied:
I remember now, you cannot have any other record types with CNAME.

So, we do need that prefix, and need to add it back in the getManagedTxt and delete functions, sorry

https://serverfault.com/questions/834320/adding-both-cname-and-txt-dns-records-for-one-subdomain

KAllan357 replied:
Fixed in https://github.com/adyanth/cloudflare-operator/pull/77/commits/062057541296aa766940186a2851b750e20c7602

adyanth reviewed (changes requested):
Comments with suggestions

KAllan357 commented:
Re-requested review. I'd like to try testing this, but I'm not 100% sure when I can get around to deploying a custom container with my changes.

adyanth commented:
The makefile should allow you to build the container. No worries though, let me check it out. Will build and check before merging in. Otherwise, looks good! Thanks again for your contribution!

KAllan357 commented:
FWIW, `make build` does not  seem to work on my machine.  Maybe a separate issue to open?

```
go: creating new go.mod: module tmp
Downloading sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0
go: added github.com/fatih/color v1.12.0
go: added github.com/go-logr/logr v0.4.0
go: added github.com/gobuffalo/flect v0.2.3
go: added github.com/gogo/protobuf v1.3.2
go: added github.com/google/go-cmp v0.5.6
go: added github.com/google/gofuzz v1.1.0
go: added github.com/inconshreveable/mousetrap v1.0.0
go: added github.com/json-iterator/go v1.1.11
go: added github.com/mattn/go-colorable v0.1.8
go: added github.com/mattn/go-isatty v0.0.12
go: added github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: added github.com/modern-go/reflect2 v1.0.1
go: added github.com/spf13/cobra v1.2.1
go: added github.com/spf13/pflag v1.0.5
go: added golang.org/x/mod v0.4.2
go: added golang.org/x/net v0.0.0-20210520170846-37e1c6afe023
go: added golang.org/x/sys v0.0.0-20210616094352-59db8d763f22
go: added golang.org/x/text v0.3.6
go: added golang.org/x/tools v0.1.5
go: added golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
go: added gopkg.in/inf.v0 v0.9.1
go: added gopkg.in/yaml.v2 v2.4.0
go: added gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
go: added k8s.io/api v0.22.2
go: added k8s.io/apiextensions-apiserver v0.22.2
go: added k8s.io/apimachinery v0.22.2
go: added k8s.io/klog/v2 v2.9.0
go: added k8s.io/utils v0.0.0-20210819203725-bdf08cb9a70a
go: added sigs.k8s.io/controller-tools v0.7.0
go: added sigs.k8s.io/structured-merge-diff/v4 v4.1.2
go: added sigs.k8s.io/yaml v1.2.0
/home/kallan/code/cloudflare-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
bash: /home/kallan/code/cloudflare-operator/bin/controller-gen: No such file or directory
make: *** [Makefile:80: generate] Error 127
```

adyanth commented:
You need to have controller-gen installed and linked to that folder it mentions for kube builder to work.

Kind of a pain especially in macOS. Brew has it iirc.

https://book.kubebuilder.io/reference/controller-gen.html

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -431,96 +332,72 @@ func (c *CloudflareAPI) GetZoneId() (string, error) {
 		c.Log.Info("Updating existing record", "fqdn", fqdn, "dnsId", dnsId)
-		method = "PUT"
-		subPath = "/" + dnsId
+		updateParams := cloudflare.UpdateDNSRecordParams{
+			Type:    "CNAME",
```
Update needs an ID

```suggestion
		updateParams := cloudflare.UpdateDNSRecordParams{
			ID:      dnsId,
			Type:    "CNAME",
```

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -561,172 +426,121 @@ func (c *CloudflareAPI) GetDNSCNameId(fqdn string) (string, error) {
+	rc := cloudflare.ZoneIdentifier(c.ValidZoneId)
+	params := cloudflare.ListDNSRecordsParams{
+		Type: "TXT",
+		Name: fqdn,
```
Needs the managed prefix here.

```suggestion
		Name: fmt.Sprintf("%s%s", TXT_PREFIX, fqdn),
```

adyanth commented on controllers/utils.go:
```diff
@@ -69,18 +70,41 @@ func getAPIDetails(ctx context.Context, c client.Client, log logr.Logger, tunnel
 	return cfAPI, cfSecret, nil
 }
+
+//getCloudflareClient returns an initialized *cloudflare.API using either an API Key + Email or an API Token
```
Go fmt complains otherwise :)

```suggestion
// getCloudflareClient returns an initialized *cloudflare.API using either an API Key + Email or an API Token
```

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -561,172 +426,121 @@ func (c *CloudflareAPI) GetDNSCNameId(fqdn string) (string, error) {
+			// TXT record exists, but not in JSON
+			c.Log.Error(err, "could not read TXT content in getting zoneId, check domain", "domain", c.Domain)
+			return records[0].ID, dnsTxtResponse, false, err
+		} else if dnsTxtResponse.TunnelId != c.ValidTunnelId {
+			// TXT record exists but not controlled by our tunnel
+			return records[0].ID, dnsTxtResponse, false, nil
```
The function should return when successful. If not here, either need an else block or pull `dnsTxtResponse` out and return at the end.

```suggestion
		} else if dnsTxtResponse.TunnelId == c.ValidTunnelId {
			// TXT record exists and controlled by our tunnel
			return records[0].ID, dnsTxtResponse, true, nil
```

adyanth commented on controllers/cloudflare_api.go:
```diff
@@ -561,172 +426,121 @@ func (c *CloudflareAPI) GetDNSCNameId(fqdn string) (string, error) {
+	if txtId != "" {
+		c.Log.Info("Updating existing TXT record", "fqdn", fqdn, "dnsId", dnsId, "txtId", txtId)
+
+		updateParams := cloudflare.UpdateDNSRecordParams{
+			Type:    "TXT",
```
This update needs an ID too.

```suggestion
		updateParams := cloudflare.UpdateDNSRecordParams{
			ID:      txtId,
			Type:    "TXT",
```

adyanth reviewed (changes requested):
Tested it out. This should be all the necessary changes.

KAllan357 commented:
Thank you! I just batched all those suggestions up and committed them.
